### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by dedobbin**
I made a setup for testing and added 2 simple tests (communication test and a time out test). I made use of [doctest](https://github.com/doctest/doctes).

Also took the opportunity to add cmake for building.

socat and pseudoterminal are used to make communication between 2 ports on the same machine possible. This is then used for the tests.  It took some messing around to get a relatively clean implementation - I'm happy with the current approach, but of course feel free to give feedback on it and other parts of the code!

### CODE REVIEW

File: tests/communication_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include "doctest.h"
+#include "pseudoterminal.h"
+#include "LinSerial.h"
+
+TEST_SUITE ("communication_tests")
+{
+    TEST_CASE ("read_single_line")
+    {
+        auto tunnel = pseudoterminal::setup();
+        LinSer::Serial writer(std::get<1>(tunnel).c_str());
```
Comment (Imrekorf) [inline tests/communication_tests.cpp]: why use a tuple and not an array?
even better would be something like a class / struct with "endpoint1" and "endpoint2" so that code is less ambigious

File: tests/helpers.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        result.push_back(token);
+        string.erase(0, pos + delimiter.length());
+    }
+    return result;
+}
+
+std::vector<std::string> get_file_names(const std::string& path)
+{
+    std::string response = execute("ls " + path);
+    // When failed, sends output to std::err so this is empty. Threat same as no files
```
Comment (Imrekorf) [inline tests/helpers.cpp]: what is meant with Threat?

File: tests/communication_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#include <iostream>
+#include <thread>
+#include <chrono>
+#include <unistd.h>
+#include <tuple>
+#include "doctest.h"
+#include "pseudoterminal.h"
+#include "LinSerial.h"
```
Comment (Imrekorf) [inline tests/communication_tests.cpp:4]: how many of these headers are actually needed?

File: tests/helpers.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#include <string>
+#include <memory>
+#include <vector>
+#include <iostream>
+#include "helpers.h"
+
+std::string execute(const std::string& cmd)
+{
+    //todo: silence output to terminal
```
Comment (Imrekorf) [inline tests/helpers.cpp]: has this been done yet?

File: tests/helpers.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+#include <algorithm>
+#include <vector>
+#include <string>
+
+std::string execute(const std::string& cmd);
+std::vector<std::string> str_split(std::string string, const std::string& delimiter);
+std::vector<std::string> get_file_names(const std::string& path);
+
+template <typename T>
+std::vector<T> find_nonintersecting_elements(std::vector<T> a, std::vector<T> b)
```
Comment (Imrekorf) [inline tests/helpers.h]: is this still used?

File: tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#define DOCTEST_CONFIG_COLORS_NONE
+#include "doctest.h"
```
Comment (Imrekorf) [inline tests/main.cpp:3]: is this file still necessary?

File: tests/helpers.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#include <string>
+#include <memory>
+#include <vector>
+#include <iostream>
+#include "helpers.h"
```
Comment (Imrekorf) [inline tests/helpers.cpp]: maybe change to something like "utility"

File: tests/pseudoterminal.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+#include <tuple>
+#include <iostream>
+#include <thread>
+#include <chrono>
+#include <fstream>
+#include <signal.h>
+#include "helpers.h"
+#include "pseudoterminal.h"
+
+const std::string OUTPUT_FILE = "socat_output.txt";
```
Comment (Imrekorf) [inline tests/pseudoterminal.cpp]: is this file ever cleaned up?

File: tests/pseudoterminal.h [hunk 9]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+namespace pseudoterminal
```
Comment (Imrekorf) [inline tests/pseudoterminal.h:5]: change to class?

File: tests/communication_tests.cpp (hunk 1 above)
Comment (dedobbin) [reply tests/communication_tests.cpp]: I refactored it into a class now.

File: tests/communication_tests.cpp (hunk 3 above)
Comment (dedobbin) [reply tests/communication_tests.cpp:4]: I did a general cleanup of the headers now

File: tests/helpers.h (hunk 5 above)
Comment (dedobbin) [reply tests/helpers.h]: Nope ! I removed it now

File: tests/pseudoterminal.h (hunk 9 above)
Comment (dedobbin) [reply tests/pseudoterminal.h:5]: Refactored it, it uses a class now

File: tests/main.cpp (hunk 6 above)
Comment (dedobbin) [reply tests/main.cpp:3]: As far as i know this provides the entry point for the test suite

File: tests/helpers.cpp (hunk 4 above)
Comment (dedobbin) [reply tests/helpers.cpp]: No, but i realized later it didn't make much sense. the caller has flexibility now to pipe output or not. Messing with that in the function itself is confusing and unnecessary.
So i removed the todo

File: tests/helpers.cpp (hunk 7 above)
Comment (dedobbin) [reply tests/helpers.cpp]: I agree helpers isn't quite the right name, but i'm always struggling with this. 

I think "utility" is also used for tools that are not part of a program/lib, but more standalone tools that are not needed for functionality.  Perhaps "common"?

I'm a bit conflicted here.

File: tests/main.cpp (hunk 6 above)
Comment (Imrekorf) [reply tests/main.cpp:3]: alright!

File: tests/helpers.cpp (hunk 7 above)
Comment (Imrekorf) [reply tests/helpers.cpp]: common would be appropriate aswell

File: tests/pseudoterminal.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+const std::string INVALID_PORT = "invalid";
+
+namespace pseudoterminal
+{
+    class ConnectedPorts
```
Comment (Imrekorf) [inline tests/pseudoterminal.h:7]: maybe change the name to pseudoterminal and remove the namespace.

File: tests/helpers.cpp (hunk 7 above)
Comment (Imrekorf) [reply tests/helpers.cpp]: helpers still not changed

**Comment by Imrekorf** [comment]
change merge request branch to development

File: tests/helpers.cpp (hunk 7 above)
Comment (dedobbin) [reply tests/helpers.cpp]: Changed it to common now

File: tests/pseudoterminal.h (hunk 10 above)
Comment (dedobbin) [reply tests/pseudoterminal.h:7]: I think this is the most natural approach. I can see a situation where logic related to pseudoterminal has to be implemented to support the setup, yet is not directly related to lifecycle of an instance of "connected ports"

Comment (Imrekorf) [reply tests/pseudoterminal.h:7]: Good point. Then I would rename the class to pseudoPort and add a terminal class that implements the connection between the ports. This way you can also future proof for multiple port connections on one terminal.

**Review by Imrekorf — CHANGES_REQUESTED** [review]
Take a look at the new style document on the development branch. Ive applied that to the current code aswell.

File: LinSerial.cpp [hunk 11]
```diff
@@ -96,438 +81,305 @@ Serial::Serial(const char* Port, const SerParam& SP, const SerTimeOut& ST){
...
-	SerialHandleMutex.lock();
+	serialHandleMutex.lock();
 	close(hSerial);
-	SerialHandleMutex.unlock();
+	serialHandleMutex.unlock();
 	hSerial = 0;
 
-	#if LINSERDEBUG >= 3
-	std::cout << "[LinSer] Serial thread closed" << std::endl;
+	#if LINSER_LOGLEVEL >= 3
```
Comment (Imrekorf) [inline LinSerial.cpp:180]: most likely a problem from the current development branch but this pre processor macro should not be used, also wrong define name if intended.

**Review by Imrekorf — CHANGES_REQUESTED** [review]

**Comment by Imrekorf** [comment]
changes moved to different branch on fork
